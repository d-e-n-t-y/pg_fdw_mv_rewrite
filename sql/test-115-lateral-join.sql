-- Anticipate message "query requires unsupported features"
set mv_rewrite.log_match_progress = 'true';

explain (VERBOSE, COSTS OFF)
select t.key, COUNT (lt.value)
  from
    test t,
    lateral (
      select key, value
        from test tt
        where t.key = tt.key
          and tt.value = (t.value::numeric + 1)::text
	) lt
  where t.value::numeric < 50
  group by t.key;

select t.key, COUNT (lt.value)
  from
    test t,
    lateral (
      select key, value
        from test tt
        where t.key = tt.key
          and tt.value = (t.value::numeric + 1)::text
	) lt
  where t.value::numeric < 50
  group by t.key;
