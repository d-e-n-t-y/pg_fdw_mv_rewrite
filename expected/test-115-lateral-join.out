-- Anticipate message "query requires unsupported features"
set mv_rewrite.log_match_progress = 'true';
explain (VERBOSE, COSTS OFF)
select t.key, COUNT (lt.value)
  from
    test t,
    lateral (
      select key, value
        from test tt
        where t.key = tt.key
          and tt.value = (t.value::numeric + 1)::text
	) lt
  where t.value::numeric < 50
  group by t.key;
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv1
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv2
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv1
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv2
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv1
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: GROUP BY list for MV and query differ in length.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite GROUP BY, but MV has no GROUP BY.
                                                    QUERY PLAN                                                     
-------------------------------------------------------------------------------------------------------------------
 Finalize GroupAggregate
   Output: t.key, count(tt.value)
   Group Key: t.key
   ->  Gather Merge
         Output: t.key, (PARTIAL count(tt.value))
         Workers Planned: 2
         ->  Partial GroupAggregate
               Output: t.key, PARTIAL count(tt.value)
               Group Key: t.key
               ->  Merge Join
                     Output: t.key, tt.value
                     Merge Cond: ((tt.key = t.key) AND (tt.value = ((((t.value)::numeric + '1'::numeric))::text)))
                     ->  Sort
                           Output: tt.value, tt.key
                           Sort Key: tt.key, tt.value
                           ->  Parallel Seq Scan on public.test tt
                                 Output: tt.value, tt.key
                     ->  Materialize
                           Output: t.key, t.value, ((((t.value)::numeric + '1'::numeric))::text)
                           ->  Sort
                                 Output: t.key, t.value, ((((t.value)::numeric + '1'::numeric))::text)
                                 Sort Key: t.key, ((((t.value)::numeric + '1'::numeric))::text)
                                 ->  Seq Scan on public.test t
                                       Output: t.key, t.value, (((t.value)::numeric + '1'::numeric))::text
                                       Filter: ((t.value)::numeric < '50'::numeric)
(25 rows)

select t.key, COUNT (lt.value)
  from
    test t,
    lateral (
      select key, value
        from test tt
        where t.key = tt.key
          and tt.value = (t.value::numeric + 1)::text
	) lt
  where t.value::numeric < 50
  group by t.key;
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv1
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv2
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv1
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv2
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv1
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: GROUP BY list for MV and query differ in length.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_from_join_clauses_are_valid_for_mv: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite GROUP BY, but MV has no GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite GROUP BY, but MV has no GROUP BY.
 key  | count 
------+-------
 key1 |     9
 key2 |    49
(2 rows)

